File: app/src/main/java/com/saizad/mvvmexample/di/auth/AuthModule.java
package com.saizad.mvvmexample.di.auth;

import com.saizad.mvvm.ActivityResult;
import com.saizad.mvvm.SaizadEasyRetrofit;
import com.saizad.mvvm.CurrentUserType;
import com.saizad.mvvm.FCMToken;
import com.saizad.mvvm.NotifyOnce;
import com.saizad.mvvmexample.api.AuthApi;

import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import io.reactivex.subjects.BehaviorSubject;
import sa.zad.easypermission.PermissionManager;

@Module
public class AuthModule {


    @Provides
    static AuthEnvironment providesAuthEnvironment(FCMToken fcmToken, AuthApi authApi, CurrentUserType currentUser, BehaviorSubject<ActivityResult<?>> navigationFragmentResult, @Named("notification") BehaviorSubject<NotifyOnce<?>> notifyOnceBehaviorSubject, PermissionManager permissionManager) {
        return new AuthEnvironment(fcmToken, authApi, navigationFragmentResult, currentUser, notifyOnceBehaviorSubject, permissionManager);
    }

    @Provides
    static AuthApi providesAuthApi(SaizadEasyRetrofit easyRetrofit) {
        return easyRetrofit.provideRetrofit().create(AuthApi.class);
    }
}

File: app/src/main/java/com/saizad/mvvmexample/di/auth/AuthModule.kt
package com.saizad.mvvmexample.di.auth

import com.saizad.mvvm.*
import com.saizad.mvvmexample.MVVMExampleCurrentUser
import com.saizad.mvvmexample.api.AuthApi
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import io.reactivex.subjects.BehaviorSubject
import sa.zad.easypermission.PermissionManager
import javax.inject.Named

@Module
@InstallIn(SingletonComponent::class)
object AuthModule {

    @Provides
    fun providesAuthEnvironment(
        authApi: AuthApi,
        fcmToken: FCMToken,
        currentUser: MVVMExampleCurrentUser,
        navigationFragmentResult: BehaviorSubject<ActivityResult<*>>,
        @Named("notification") notifyOnceBehaviorSubject: BehaviorSubject<NotifyOnce<*>>,
        permissionManager: PermissionManager
    ): AuthEnvironment {
        return AuthEnvironment(
            authApi,
            fcmToken,
            navigationFragmentResult,
            currentUser,
            notifyOnceBehaviorSubject,
            permissionManager
        )
    }


    @Provides
    fun providesAuthApi(easyRetrofit: SaizadEasyRetrofit): AuthApi {
        return easyRetrofit.provideRetrofit().create(AuthApi::class.java)
    }
}